### PR TITLE
sd-boot: Use UEFI provided CRC32

### PR DESCRIPTION
**Pull request by medhefgo**

### CODE REVIEW

File: src/boot/efi/boot.c [hunk 1]
```diff
@@ -2204,7 +2202,7 @@ static VOID config_load_xbootldr(
                                 continue;
 
                         /* Calculate CRC of entries array, too */
-                        c = ~crc32(UINT32_MAX, entries, sz);
+                        r = BS->CalculateCrc32(&entries, sz, &c);
```
Comment (yuwata) [inline src/boot/efi/boot.c]: Not familiar with that, and I cannot find the manual of the function, but is it not necessary to check the return value `r`?
If not necessary, then please voidify it: `(void) BS->CalculateCrc32(...);`

Comment (medhefgo) [reply src/boot/efi/boot.c]: You're right, fixed it.

**Comment by yuwata** [comment]
Not familiar with the function. If `CalculateCrc32()` is always supported, then it looks superficially good for me.

**Comment by medhefgo** [comment]
The oldest UEFI spec I can pull from their site is from 2006 and it has it. If it turns out it's unsupported somewhere, we can also fall back on gnu-efi provided crc32.

File: src/boot/efi/boot.c [hunk 2]
```diff
@@ -2170,12 +2169,11 @@ static VOID config_load_xbootldr(
                                 continue;
 
                         /* Calculate CRC check */
-                        c = ~crc32_exclude_offset(UINT32_MAX,
-                                                  (const UINT8*) &gpt_header_buffer,
-                                                  h->Header.HeaderSize,
-                                                  OFFSETOF(EFI_PARTITION_TABLE_HEADER, Header.CRC32),
-                                                  sizeof(h->Header.CRC32));
-                        if (c != h->Header.CRC32)
+                        c_expect = h->Header.CRC32;
```
Comment (poettering) [inline src/boot/efi/boot.c]: wouldn't a better name for this var be "crc32_saved" or just "saved", or so?

File: src/boot/efi/boot.c [hunk 3]
```diff
@@ -2170,12 +2169,11 @@ static VOID config_load_xbootldr(
...
+                        h->Header.CRC32 = 0;
+                        r = BS->CalculateCrc32(&gpt_header_buffer, h->Header.HeaderSize, &c);
```
Comment (poettering) [inline src/boot/efi/boot.c]: so this function already inverts the bits?

**Review by poettering — APPROVED** [review]
i figure you tested this? and the bitwise inversion is already done by that call?

**Comment by medhefgo** [comment]
Yes, I verified (within file_read to get some easy target to compare) to make they get the same output, and yes, it already does the xor.

**Comment by poettering** [comment]
xor? i meant the bitwise inversion (`~`)

**Comment by medhefgo** [comment]
The comment in crc32() says xor (with presumably all bits set to 1, so a not :D) . Either way, the way the call is now is correct.

**Comment by medhefgo** [comment]
You make that sounds like I know my binary (I do not). :D

**Comment by bluca** [comment]
Test failure unrelated:

```
=== Failed test log ===
--- test-journal-flush begin ---
journal_file_copy_entry failed: Bad message
Assertion 'r >= 0' failed at src/libsystemd/sd-journal/test-journal-flush.c:52, function main(). Aborting.
```

**Pull request merged**